client.go

```diff
@@ -0,0 +1,289 @@
+	}
+
+	return &ret
+}
+
+type Client struct {
+	cfg   *settings
+	creds Credentials
+}
+
+func runWithRetry[TREQ any, TRESP commonResponseSetter](
+	ctx context.Context,
+	logger *logs.Logger,
+	req *request[TREQ],
+) (
+	resp *response[TRESP],
+	_ error,
+) {
+	reterr := retry.ExpBackoff(ctx, logger, retryFirsDelay, retryMaxDelay,
+		retryFactor, retryMaxAttempts, func() error {
+			var err error
+			resp, err = runOnce[TREQ, TRESP](ctx, logger, req)
+			return err
+		})
+
+	return resp, reterr
+}
+
+// runOnce sends an HTTP request, parses and returns the response.
+// Permanent errors are wrapped with retry.PermanentError. Any error returned
+// from the server is wrapped with HTTPError. If the error is a valid
+// CloudFlare error, it is also wrapped with CloudFlareError.
+func runOnce[TREQ any, TRESP commonResponseSetter](
+	ctx context.Context,
+	logger *logs.Logger,
+	treq *request[TREQ],
+) (
+	tresp *response[TRESP],
+	err error,
+) {
+	if err = treq.client.cfg.ratelim.Wait(ctx); err != nil {
+		return
+	}
+
+	// request body
+	var reqBody []byte
+	if treq.body != nil {
+		reqBody, err = json.Marshal(treq.body)
+		if err != nil {
+			err = retry.PermanentError{Cause: err}
+			return
+		}
+	}
+
+	// request
+	req, err := http.NewRequestWithContext(ctx, treq.method, requestURL(treq),
+		bytes.NewReader(reqBody))
+	if err != nil {
+		err = retry.PermanentError{Cause: err}
+		return
+	}
+
+	// headers
+	mergeHeaders(req.Header, treq.headers)
+
+	// credentials
+	err = treq.client.creds.configure(
+		logger.SubLogger(logs.WithPrefix("Authorization")),
+		req)
+	if err != nil {
+		return // allow retry
+	}
+
+	// send the request
+	resp, err := treq.client.cfg.httpClient.Do(req)
+	if err != nil {
+		// errors from Do() may be permanent or not, it is not possible to
+		// determine precisely
+		return // allow retry
+	}
+
+	// handle response
+	if resp.StatusCode >= 400 {
+		err = handleErrorResponse(resp, logger)
+		logFullRequestError(logger, treq, reqBody, err)
+		return
+	}
+
+	tresp, err = handleSuccessResponse[TRESP](resp, logger)
+	if err != nil {
+		logFullRequestError(logger, treq, reqBody, err)
+		return
+	}
+
+	if treq.client.cfg.logSuccess {
+		logFullHTTPRequestSuccess(logger, treq, reqBody, tresp)
+	}
+
+	return tresp, err
+}
+
+func handleSuccessResponse[TRESP commonResponseSetter](httpResp *http.Response, logger *logs.Logger) (
+	*response[TRESP],
+	error,
+) {
+	var ret response[TRESP]
+
+	ret.code = httpResp.StatusCode
+	ret.headers = httpResp.Header
+
+	var err error
+	ret.rawBody, err = io.ReadAll(httpResp.Body)
+	if err != nil {
+		// allow retry
+		return nil, errors.Join(err, HTTPError{
+			Code:    httpResp.StatusCode,
+			RawBody: ret.rawBody,
+			Headers: ret.headers,
+		})
+	}
+
+	err = json.Unmarshal(ret.rawBody, &ret.body)
+	if err != nil {
+		// allow retry
+		return nil, errors.Join(err, HTTPError{
+			Code:    httpResp.StatusCode,
+			RawBody: ret.rawBody,
+			Headers: ret.headers,
+		})
+	}
+
+	return &ret, nil
```

Comment:
The response struct is internal, it will not be included on the response from any caller from an external package, like the consumers of the library. The response body is released with it, so there is no real difference to how much memory is used. And then I would have to add two parameters to `logFullHTTPRequestSuccess` to send the response header and body. I really think it would make the code more complicated and won't bring benefits.